kolibri/core/exams/serializers.py

```diff
@@ -164,45 +198,95 @@ def _create_exam_assignment(self, **params):
         )
 
     def update(self, instance, validated_data):  # noqa
-        # Update the scalar fields
-        instance.title = validated_data.get("title", instance.title)
-        instance.active = validated_data.get("active", instance.active)
-        instance.archive = validated_data.get("archive", instance.archive)
-        instance.date_activated = validated_data.get(
-            "date_activated", instance.date_activated
-        )
-        instance.date_archived = validated_data.get(
-            "date_archived", instance.date_archived
-        )
+        # Out of an abundance of caution, handle the saving of the new instance and deletion of the old instance
+        # in a transaction, so that if an error occurs in either, we don't end up with a mismatched state
+        # to make this simpler, we wrap the whole update in a transaction
+        with transaction.atomic():
+            instance_is_draft = isinstance(instance, DraftExam)
+            new_draft_value = validated_data.pop("draft", instance_is_draft)
+            if not instance_is_draft and new_draft_value:
+                raise ValidationError(
+                    "Cannot change an Exam to a DraftExam", code=error_constants.INVALID
+                )
+            elif instance_is_draft and not new_draft_value:
+                if "assignments" not in validated_data:
+                    validated_data["assignments"] = Collection.objects.filter(
+                        id__in=instance.assignments
+                    )
+                if "learner_ids" not in validated_data:
+                    validated_data["learner_ids"] = FacilityUser.objects.filter(
+                        id__in=instance.learner_ids
+                    )
+                new_instance = instance.to_exam()
+                new_instance.save()
+                instance.delete()
+                instance = new_instance
+                instance_is_draft = False
```

Comment:
Did you consider doing this in `to_exam()` if the exam saves successfully? When I read that code I was thinking maybe auto-deletion should be enforced in the model. Nothing blocking just something that came to mind.